File: dashboard/src/lib/api/scope.ts
// Environment scoping for the axios request interceptor in `client.ts`.
//
// Split into its own module for the same reason auth has its own bridge in
// `client.ts`: `client.ts` must not import a store directly, because
// `sessionStore` (transitively, via `../api/{orgs,apps,environments}`)
// imports `client.ts` itself — a module-level `import { sessionStore }` here
// would recreate that `store -> api -> client -> store` cycle one hop later
// (`client.ts -> scope.ts -> session.svelte.ts -> apps.ts -> client.ts`).
//
// So this module takes the same approach as `configureAuthBridge`: it holds
// no store import, only a bridge interface the store wires itself into once,
// at construction (see `SessionStore`'s constructor in `session.svelte.ts`).

export interface ScopeBridge {
  /**
   * The currently selected environment id, or `null` for "all environments".
   * The literal string `'none'` means "unattributed" and is a real id as far
   * as this module is concerned — it is passed straight through to the wire.
   */
  getCurrentEnvironmentId(): string | null;
}

const noopScopeBridge: ScopeBridge = {
  getCurrentEnvironmentId: () => null,
};

let bridge: ScopeBridge = noopScopeBridge;

export function configureScopeBridge(next: ScopeBridge): void {
  bridge = next;
}

/** Reads the current environment id through the bridge. `null` = all environments. */
export function currentEnvironmentId(): string | null {
  return bridge.getCurrentEnvironmentId();
}

// ---------------------------------------------------------------------------
// The rule is opt-IN, not opt-out.
//
// This module used to work the other way around: scope everything except a
// short list of URL substrings (`/environments`, `/first-event`, `/funnels`,
// `/artifacts`, plus a bare-`/v1/apps/{id}` regex). An opt-out list fails
// *open* — anything the list's author forgot, or any route family added
// later, gets `environment_id` attached by default. That is exactly what
// happened: a grep of `backend/bins/sauron-api/src/routes/` for
// `reject_environment_id` (the backend's hard-`400`-on-any-value guard, see
// `routes/scope.rs`) turns up five call sites, and the opt-out list only
// accounted for two of them (`funnels.rs`, `artifacts.rs`):
//
//   - `admin.rs`         -> GET  /v1/admin/storage
//   - `artifacts.rs`     -> already covered (app-scoped, see below)
//   - `funnels.rs`       -> already covered (app-scoped, see below)
//   - `monitors.rs`      -> /v1/projects/{id}/monitors,
//                           /v1/monitors/{id}, /v1/monitors/{id}/checks,
//                           /v1/monitors/{id}/incidents
//   - `notifications.rs` -> /v1/orgs/{id}/notification-channels,
//                           /v1/notification-channels/{id}[/test],
//                           /v1/orgs/{id}/alert-rules, /v1/alert-rules/{id},
//                           /v1/orgs/{id}/alert-events, /v1/alert-meta
//
// None of the monitors/notifications/admin URLs matched the opt-out
// substrings, so every request on Settings > Alerting, Uptime > Monitors and
// Admin > Storage was getting `environment_id` attached and rejected with a
// 400 — not an edge case, since `sessionStore.currentEnvId` is non-null by
// default (`resolveCurrentEnvironment` auto-selects the app's default
// environment).
//
// The fix inverts the default: only telemetry reads under
// `/v1/apps/{app_id}/...` are eligible for scoping at all. Everything else —
// `/v1/monitors/...`, `/v1/projects/...`, `/v1/orgs/...`, `/v1/admin/...`,
// `/v1/auth/...`, `/v1/environments/...` (and any route family added after
// this comment was written) — is unscoped *by construction*, with no list to
// forget to update. A new non-telemetry route is safe by default; a new
// telemetry route under `/v1/apps/{app_id}/...` is scoped by default and
// must be added to the exclusion list below only if the backend actually
// rejects the parameter (verify against `reject_environment_id` call sites,
// don't guess from the URL shape).
const APP_SCOPED_URL = /^\/v1\/apps\/[^/]+(?:\/.*)?$/;

// Within `/v1/apps/{app_id}/...`, these sub-resources are app *configuration*
// rather than telemetry, and the backend calls `reject_environment_id` (or
// `reject_environment_id_with_message` — same grep target, its name is a
// prefix of the former) on all of them. This array is load-bearing in BOTH
// directions now, not just backend-to-dashboard: Task 14's
// `backend/bins/sauron-api/tests/http_env_scoping.rs` reads this exact array
// out of this file's source (the same way `permissions.test.ts` reads
// `rbac.rs` in the other direction) and asserts it equals the set of
// app-scoped GETs that actually 400 on a valid `environment_id` when driven
// over HTTP — so an entry added here without a matching backend
// `reject_environment_id*` call, or a backend rejection added without a
// matching entry here, now fails a test instead of drifting silently. Each
// entry below is one of the `reject_environment_id*` call-site groups,
// verified directly against `backend/bins/sauron-api/src/routes/` and
// `main.rs`'s route table:
//
//  - `/v1/apps/{id}` bare (`apps::get_app`) — app metadata (name,
//    ingest_enabled) has no environment dimension. Calls
//    `reject_environment_id`; a 400 without this exclusion.
//
//  - `/v1/apps/{id}/environments` (`environments::list_environments`) — this
//    IS the environment list; scoping the request that enumerates
//    environments to one of them is circular. Calls `reject_environment_id`;
//    a 400 without this exclusion.
//
//  - `/v1/apps/{id}/funnels` (`funnels::list_saved` / `create_saved` /
//    `update_saved` / `delete_saved`) — saved funnel *definitions* (plural,
//    app-wide config: name/description/step list). Calls
//    `reject_environment_id`; a 400 without this exclusion.
//    NOT the same route as `/v1/apps/{id}/funnel` (singular, `POST`,
//    `funnels::compute`) — that one computes live counts for a chosen date
//    range and calls `read_scope`, i.e. it IS telemetry and must stay
//    scoped. The two differ by exactly the trailing "s", so this is matched
//    as a literal path segment, not a word-in-URL substring, to avoid a
//    plural match accidentally swallowing the singular. Verified against the
//    route table in `main.rs` rather than pattern-matched from the word.
//
//  - `/v1/apps/{id}/artifacts` (`artifacts::list` / `upload` / `delete`) —
//    symbol artifacts are app-wide config, not telemetry. Calls
//    `reject_environment_id`; a 400 without this exclusion.
//
//  - `/v1/apps/{id}/errors/timeseries`, `/v1/apps/{id}/events/timeseries`,
//    `/v1/apps/{id}/transactions/timeseries` (`analytics::error_timeseries` /
//    `event_timeseries` / `transaction_timeseries`) — cross-tier reads that
//    route across hot Postgres and cold Parquet; cold storage is not
//    partitioned by environment yet, so these reject any `environment_id` at
//    all rather than scoping only the hot half. Each calls
//    `reject_environment_id_with_message` (a `reject_environment_id*` call
//    site like every other entry above — the grep below finds it because its
//    name is a prefix of `reject_environment_id`), which preserves their own
//    reason string instead of the generic one. Previously these three
//    hand-rolled an inline `raw_environment_id(..).is_some()` check instead of
//    calling through `routes::scope` at all, which is exactly what made the
//    grep below miss them — fixed at the source rather than patched here only.
//
// `get_app` and `list_environments` used to parse no `environment_id` at all
// (an extra query param was silently dropped, not rejected) — Task 14's
// router-enumeration test caught that as the exact defect class this whole
// module exists to prevent (a 200 on a malformed `environment_id` looks like
// a filter was applied and wasn't), so both now call `reject_environment_id`
// too. This array's comment used to call those two "precautionary rather
// than required"; that is no longer true, and the wording above has been
// corrected rather than left stale.
//
//  - `/v1/apps/{id}/inspector/policy` (`inspector::effective_policy`) — the
//    PII inspector is APP-scoped. Findings carry their own environment
//    dimension inside the payload (`env_scope` plus `environment_id`), and
//    MASKING cannot be limited to one environment at all: the pipeline
//    enforcer keys on `app_id` alone, and a policy that masks in prod but not
//    staging is a footgun that produces exactly the leak the feature exists to
//    prevent. It calls `reject_environment_id_with_message` with that reason.
//
//  - `/v1/apps/{id}/inspector/mask-actions`
//    (`inspector::list_app_mask_actions`) and
//    `/v1/apps/{id}/inspector/masked-keys`
//    (`inspector::list_app_masked_keys`) — the same app-scoped rule, the same
//    message, the same helper. A mask action's `targets` name a table, a
//    column and a json path and nothing else, and `inspector_masked_keys` has
//    no environment column at all, so there is no environment dimension to
//    narrow on: accepting the parameter would report a filter that was never
//    applied. Added in the same change that MOUNTED both routes, because the
//    Rust contract test enumerates `main.rs` and an entry added here ahead of
//    its route fails that test exactly as loudly as a missing one does.
//    `POST /v1/apps/{id}/inspector/mask-preview` rejects `environment_id` too
//    and is deliberately NOT listed HERE: `http_env_scoping.rs`'s
//    `app_scoped_get_route_templates()` only collects `.route(...)` calls
//    containing `get(`, and
//    `the_backend_rejection_set_matches_the_dashboard_exclusion_list` asserts
//    that collected set EQUALS this array. A POST-only path here would sit in
//    `expected` and could never appear in `rejecting`, so the contract test
//    would fail on a perfectly correct handler. Every entry that IS listed has
//    a GET. It is excluded from SCOPING all the same, one array down — see
//    `NON_GET_BACKEND_REJECTIONS`.
const BACKEND_REJECTS_ENVIRONMENT_ID: RegExp[] = [
  /^\/v1\/apps\/[^/]+\/?(?:\?.*)?$/,
  /^\/v1\/apps\/[^/]+\/environments(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/funnels(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/artifacts(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/errors\/timeseries(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/events\/timeseries(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/transactions\/timeseries(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/inspector\/policy(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/inspector\/mask-actions(?:[/?].*)?$/,
  /^\/v1\/apps\/[^/]+\/inspector\/masked-keys(?:[/?].*)?$/,
];

// `/v1/apps/{id}/first-event` (`apps::first_event`) is different in kind from
// the array above: the backend *does* support scoping it for real
// (`read_scope`, not `reject_environment_id` — it would happily narrow on a
// valid `environment_id`). It is excluded here anyway, for a reason specific
// to the ONE call site that hits it: `Onboarding.svelte` polls a
// just-created app by that app's own id, not necessarily
// `sessionStore.currentAppId` — the store's `currentEnvId` can belong to a
// different app entirely. Attaching it blindly would scope the poll to an
// environment that doesn't belong to the app being onboarded, silently
// under-reporting "no events yet" forever.
//
// Kept as a SEPARATE array from `BACKEND_REJECTS_ENVIRONMENT_ID` on purpose:
// that array is checked against the backend's actual rejection set by
// `http_env_scoping.rs` (see the comment above it), and `first-event` is not
// part of that correspondence — folding it in would make the Rust test
// demand the backend reject a route it correctly narrows on.
const UI_ONLY_EXCLUSIONS: RegExp[] = [/^\/v1\/apps\/[^/]+\/first-event(?:[/?].*)?$/];

// App-scoped routes the backend rejects `environment_id` on that have NO `get(`
// handler, so they cannot live in `BACKEND_REJECTS_ENVIRONMENT_ID` without
// failing `http_env_scoping.rs`'s equality assertion (see the note at the end of
// that array's comment). They still have to be excluded from scoping, because
// the exclusion list is what `shouldScopeUrl` reads — being absent from it is
// not "unchecked", it is "scoped".
//
// Measured, before this array existed: with an environment selected (the
// default — `resolveCurrentEnvironment` auto-selects one), the MaskDialog's
// first call went out as
// `POST /v1/apps/{id}/inspector/mask-preview?environment_id=…` and came back
// 400 "the inspector is app-scoped; masking cannot be limited to one
// environment". The dialog then sat on "Counting affected rows…" forever with
// an empty app slug, so masking was unreachable from the UI entirely.
const NON_GET_BACKEND_REJECTIONS: RegExp[] = [
  /^\/v1\/apps\/[^/]+\/inspector\/mask-preview(?:[/?].*)?$/,
];

const APP_CONFIG_SUBPATHS: RegExp[] = [
  ...BACKEND_REJECTS_ENVIRONMENT_ID,
  ...UI_ONLY_EXCLUSIONS,
  ...NON_GET_BACKEND_REJECTIONS,
];

/** True if `url` is a read that environment scoping applies to at all. */
export function shouldScopeUrl(url: string | undefined): boolean {
  if (!url) return false;
  if (!APP_SCOPED_URL.test(url)) return false;
  return !APP_CONFIG_SUBPATHS.some((re) => re.test(url));
}

/**
 * The query params the interceptor should merge onto a request, or
 * `undefined` when it should add nothing at all.
 *
 * Two independent reasons to add nothing, both load-bearing:
 *  - `url` is not an app-scoped telemetry read (see above).
 *  - `envId` is `null` ("all environments") — the parameter must be omitted
 *    entirely rather than sent empty. The backend treats a *present* but
 *    empty `?environment_id=` as a `400`, not "all" (that silent-widening
 *    bug was a Critical in an earlier review); sending nothing is the only
 *    correct way to ask for every environment.
 */
export function computeScopeParams(
  url: string | undefined,
  envId: string | null,
): Record<string, string> | undefined {
  if (!shouldScopeUrl(url)) return undefined;
  if (envId === null) return undefined;
  return { environment_id: envId };
}

// ---------------------------------------------------------------------------
// Project-scoped routes.
//
// `APP_SCOPED_URL` above only matches `/v1/apps/...`, so nothing under
// `/v1/projects/...` is ever scoped by the interceptor — that part is safe by
// construction and this array changes no behaviour here. What it does is close
// the same two-directional gap for a NEW route family: the active-users
// endpoints are the first telemetry reads outside `/v1/apps/{id}/…`, so they
// sit outside the only mechanised check that a telemetry GET resolves
// environment scoping rather than accepting-and-ignoring it.
// `backend/bins/sauron-api/tests/http_env_scoping.rs` reads THIS array's
// literal source and asserts it equals the set of project-scoped GETs that
// actually 400 on a valid `environment_id`.
//
// Only the rejecting routes belong here. `/v1/projects/{id}` and
// `/v1/projects/{id}/apps` neither narrow nor reject — they are ordinary
// configuration reads with no environment dimension and no `Query` field for
// one — so listing them would make the Rust test demand a rejection the
// backend does not perform.
export const PROJECT_SCOPED_REJECTS_ENVIRONMENT_ID: RegExp[] = [
  /^\/v1\/projects\/[^/]+\/active-users(?:[/?].*)?$/,
  /^\/v1\/projects\/[^/]+\/active-users\.csv(?:[/?].*)?$/,
  /^\/v1\/projects\/[^/]+\/environments(?:[/?].*)?$/,
  /^\/v1\/projects\/[^/]+\/monitors(?:[/?].*)?$/,
];

File: backend/bins/sauron-api/src/routes/scope.rs
//! HTTP-layer parsing of the `environment_id` query parameter into
//! `sauron_db::scope::ReadScope`/`EnvFilter`, and — via
//! [`authorized_read_scope`] — the single call every environment-scoped
//! handler makes to turn that parameter into an authorized read.
//!
//! | value | meaning | reach required |
//! |---|---|---|
//! | absent | every environment the caller may read (auto-narrowed to a `Subset` for a partial-reach caller) | any reach on the app |
//! | `?environment_id=<uuid>` | that environment only | a grant reaching that specific environment (app-wide or that env's own) |
//! | `?environment_id=none` | rows with `environment_id IS NULL` | app-wide reach — unattributed rows belong to no single environment |
//! | anything else (including empty, i.e. `?environment_id=`) | `400` — **never** a silent fallback to "all" | n/a |
//!
//! A malformed value must be a `400`, not a silent fallback to `All`: falling
//! back would show the caller MORE data than they asked for, which is the
//! wrong direction to fail on a scoping parameter. The reach column is what
//! Slice 3 added: `environment_id` is now an access boundary enforced by
//! [`sauron_auth::authorize_env_read`], not just a display filter — see
//! [`authorized_read_scope`].
//!
//! ## The extractor trap, and why callers use [`raw_environment_id`]
//!
//! `?environment_id=` — the parameter present with an **empty** value — is
//! wire-indistinguishable from "absent" to some deserializers but not others:
//!
//! - `axum::extract::Query` (`serde_urlencoded`) deserializes an empty value
//!   into `Option<String>` as `Some("")`, which [`parse_env`] correctly
//!   rejects.
//! - `axum_extra::extract::Query` (`serde_html_form`, needed elsewhere in this
//!   crate for repeated-key `Vec<String>` fields like `filter=a&filter=b`,
//!   which `serde_urlencoded` cannot deserialize) treats an empty value as
//!   equivalent to the key being *absent* for `Option<T>`, producing `None`.
//!   Fed into [`parse_env`], `None` means `EnvFilter::All` — so a caller who
//!   typed `?environment_id=` (e.g. an empty store value interpolated into a
//!   URL) silently got every environment instead of a `400`.
//!
//! This is exactly backwards for a scoping parameter, and it depended on
//! which extractor a route handler happened to import — nothing about the
//! bug was visible from `parse_env` itself, which is correct in isolation
//! either way. The fix is to stop asking a `Query<T>` deserializer at all:
//! [`raw_environment_id`] reads `environment_id` directly out of the raw
//! query string (via `axum::extract::RawQuery`, upstream of any `Query`
//! codec), so presence/absence/emptiness is decided the same way regardless
//! of which extractor a handler's *other* query parameters go through.
//! [`authorized_read_scope`] is built on that source; every environment-scoped
//! handler calls it (passing `RawQuery`'s inner `Option<&str>`) rather than
//! feeding a `Query<T>`-deserialized `Option<String>` field into `parse_env`
//! by hand, so a future route gets this right by construction rather than by
//! remembering to.

use sauron_db::scope::{EnvFilter, ReadScope};
use sauron_db::AsyncPgConnection;
use uuid::Uuid;

use crate::error::ApiError;

/// Read `environment_id`'s raw value directly out of the query string,
/// distinguishing "the key is absent" from "the key is present with an empty
/// value" — a distinction a `Query<T>`-deserialized `Option<String>` field
/// loses (inconsistently across codecs; see the module docs above).
///
/// Returns `None` when `environment_id` does not appear at all. Returns
/// `Some(value)` (percent-decoded) when it does — `Some("")` for a bare
/// `?environment_id` or `?environment_id=`. That `Some("")` is what makes
/// [`parse_env`] reject it with a `400` instead of a `Query<T>` extractor
/// silently having already turned it into `None`.
pub fn raw_environment_id(raw_query: Option<&str>) -> Option<String> {
    let raw_query = raw_query?;
    form_urlencoded::parse(raw_query.as_bytes())
        .find(|(k, _)| k == "environment_id")
        .map(|(_, v)| v.into_owned())
}

/// Parse the `environment_id` query parameter into an [`EnvFilter`].
///
/// `raw` is the extracted `Option<String>` field's `.as_deref()`, i.e. `None`
/// when the parameter was absent from the query string.
///
/// **Syntax only.** A `One(uuid)` is only validated as a well-formed UUID —
/// this does not check that the environment exists, or that it belongs to the
/// `app_id` the caller is also passing. The existence + app-ownership check
/// this doc comment used to ask Slice 3 for now happens in
/// `sauron_auth::rbac::resolve_env_filter` (via [`authorized_read_scope`]):
/// a `One(uuid)` naming an environment that doesn't exist, or belongs to a
/// different app, is refused with a `403`, not silently narrowed to zero rows.
pub fn parse_env(raw: Option<&str>) -> Result<EnvFilter, ApiError> {
    match raw {
        None => Ok(EnvFilter::All),
        Some("none") => Ok(EnvFilter::Unattributed),
        Some(s) => Uuid::parse_str(s).map(EnvFilter::One).map_err(|_| {
            ApiError::BadRequest(format!(
                "invalid environment_id: {s:?} (expected a UUID or \"none\")"
            ))
        }),
    }
}

/// Authorize an environment-scoped read and produce its `ReadScope` in one
/// call, sourcing `environment_id` from the raw query string.
///
/// This supersedes calling `authorize_app` and a hand-built `ReadScope`
/// separately. Both orderings of that pair were correct, but only if both
/// were present — and the whole history of this feature is defects where two
/// things that had to agree were maintained by hand (this module's own
/// `?environment_id=` extractor-trap regression, documented above, is one of
/// them). One call cannot half-happen.
///
/// `raw_query` is `axum::extract::RawQuery`'s inner `Option<&str>` — every
/// caller must extract it via `RawQuery`, not read `environment_id` off a
/// `Query<T>`-deserialized struct field, for the same reason
/// [`raw_environment_id`] exists (see the module docs' "extractor trap"
/// section).
pub async fn authorized_read_scope(
    conn: &mut AsyncPgConnection,
    user_id: Uuid,
    app_id: Uuid,
    permission: &str,
    raw_query: Option<&str>,
) -> Result<ReadScope, ApiError> {
    let requested = parse_env(raw_environment_id(raw_query).as_deref())?;
    let scope =
        sauron_auth::authorize_env_read(conn, user_id, app_id, permission, requested).await?;
    Ok(scope)
}

/// [`authorized_read_scope`], plus the caller's effective permission set at the
/// **resolved** scope — for handlers that gate a second capability on top of the
/// read itself.
///
/// Every current caller gates the same one: `source:read` over the
/// de-obfuscated source lines in an `ErrorEvent`'s `stacktrace_symbolicated`.
/// As of this comment there are six: `issues::detail`, `issues::events`,
/// `sessions::detail`, `devices::detail`, `screens::detail`,
/// `analytics::person` — `grep -n authorized_read_scope_with_perms
/// src/routes/*.rs` finds each of them. That grep returns NINE lines, not eight:
/// the six call sites, this file's doc and definition lines, and `mod.rs:74`,
/// which is prose in a comment rather than a call site. The last four were added when the gate turned out to be
/// enforced only on the issues pair while those four returned whole event rows,
/// context lines included, off `event:read` alone.
///
/// Use this instead of pairing [`authorized_read_scope`] with a separate
/// permission lookup. The separate lookup it replaces (`super::authorize_app_perms`,
/// now deleted) resolved permissions at `env: None`, which no environment-scoped
/// grant can ever satisfy — so those handlers `403`'d an env-scoped caller even
/// on their own environment. See `sauron_auth::authorize_env_read_with_perms`
/// for the full account; the ordering guarantee is that authorization happens
/// first, and the permission set is computed only after the scope is resolved.
pub async fn authorized_read_scope_with_perms(
    conn: &mut AsyncPgConnection,
    user_id: Uuid,
    app_id: Uuid,
    permission: &str,
    raw_query: Option<&str>,
) -> Result<(ReadScope, std::collections::HashSet<String>), ApiError> {
    let requested = parse_env(raw_environment_id(raw_query).as_deref())?;
    let resolved =
        sauron_auth::authorize_env_read_with_perms(conn, user_id, app_id, permission, requested)
            .await?;
    Ok(resolved)
}

/// Reject `environment_id` outright rather than silently ignoring it.
///
/// For endpoints whose underlying resource has no environment dimension at
/// all — project-scoped monitors, org-scoped alert/notification config, saved
/// funnel *definitions* (as opposed to `funnels::compute`, which reads live
/// signal data and *is* scoped), symbol artifacts, and the admin storage
/// report. Accepting the parameter there and doing nothing with it is the
/// same class of bug as falling back to `All` on a scopeable endpoint: the
/// caller believes a filter was applied, and it silently wasn't.
pub fn reject_environment_id(raw: Option<&str>) -> Result<(), ApiError> {
    reject_environment_id_with_message(raw, "environment_id is not supported on this endpoint")
}

/// [`reject_environment_id`], but with a caller-supplied message instead of
/// the generic one — for an endpoint that wants to name its own specific
/// limitation (e.g. the cross-tier timeseries handlers in `analytics.rs`:
/// "cold storage is not partitioned by environment") rather than the generic
/// "not supported on this endpoint".
///
/// Still a `reject_environment_id*` call site as far as
/// `dashboard/src/lib/api/scope.ts`'s reconciliation grep is concerned — its
/// name is a prefix of `reject_environment_id`, so `grep reject_environment_id`
/// finds it too. That is deliberate: it is what lets an endpoint with a
/// bespoke rejection reason go through this module instead of hand-rolling an
/// inline `raw_environment_id(..).is_some()` check that the grep cannot see
/// (see the three `analytics.rs` timeseries handlers, and the review that
/// caught them not doing this).
pub fn reject_environment_id_with_message(
    raw: Option<&str>,
    message: &str,
) -> Result<(), ApiError> {
    if raw.is_some() {
        return Err(ApiError::BadRequest(message.into()));
    }
    Ok(())
}

/// Shared query shape for endpoints that only need to *reject* an
/// `environment_id`, with no query parameters of their own worth a bespoke
/// struct.
#[derive(Debug, serde::Deserialize)]
pub struct RejectEnvQuery {
    pub environment_id: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn absent_means_all() {
        assert_eq!(parse_env(None).unwrap(), EnvFilter::All);
    }

    #[test]
    fn none_means_unattributed() {
        assert_eq!(parse_env(Some("none")).unwrap(), EnvFilter::Unattributed);
    }

    #[test]
    fn a_uuid_selects_one() {
        let id = Uuid::from_u128(42);
        assert_eq!(
            parse_env(Some(&id.to_string())).unwrap(),
            EnvFilter::One(id)
        );
    }

    /// A malformed value is a 400, NOT a silent fallback to All. Falling back
    /// would show the caller MORE data than they asked for, which is the
    /// wrong direction to fail on a scoping parameter.
    ///
    /// **This test does NOT cover the wire-level empty-value bug** that
    /// shipped in `analytics.rs`/`issues.rs` (see the module docs' "extractor
    /// trap" section): `parse_env(Some(""))` is an input this function can
    /// never actually receive from those two files' real `Query` extractor,
    /// because `axum_extra::extract::Query`'s codec (`serde_html_form`) never
    /// hands `parse_env` a `Some("")` for `?environment_id=` — it collapses
    /// that case to `None` before `parse_env` is ever called, so `parse_env`
    /// was always correct and the bug was entirely upstream of it. A unit
    /// test at this level, however thorough, cannot see a defect that lives
    /// in which extractor a handler imports; only a test that goes through
    /// the real axum router (`tests/http_env_scoping.rs`) can. Do not read
    /// this test passing as evidence that the wire-level case is covered.
    #[test]
    fn malformed_is_rejected_not_widened() {
        assert!(parse_env(Some("not-a-uuid")).is_err());
        assert!(parse_env(Some("")).is_err());
    }

    #[test]
    fn reject_environment_id_passes_when_absent() {
        assert!(reject_environment_id(None).is_ok());
    }

    #[test]
    fn reject_environment_id_rejects_any_value_even_a_valid_one() {
        assert!(reject_environment_id(Some("none")).is_err());
        let id = Uuid::from_u128(9);
        assert!(reject_environment_id(Some(&id.to_string())).is_err());
    }

    #[test]
    fn reject_environment_id_with_message_passes_when_absent() {
        assert!(reject_environment_id_with_message(None, "custom reason").is_ok());
    }

    #[test]
    fn reject_environment_id_with_message_carries_the_custom_reason() {
        let err = reject_environment_id_with_message(Some("none"), "cold storage reason")
            .expect_err("must reject");
        assert!(
            format!("{err:?}").contains("cold storage reason"),
            "the custom message must reach the error, not the generic one: {err:?}"
        );
    }

    // --- raw_environment_id: the part a Query<T> deserializer can't do -----

    #[test]
    fn raw_absent_query_string_is_none() {
        assert_eq!(raw_environment_id(None), None);
    }

    #[test]
    fn raw_key_missing_from_query_string_is_none() {
        assert_eq!(raw_environment_id(Some("since_days=7&limit=20")), None);
    }

    /// The whole point: present-but-empty must come back `Some("")`, NOT
    /// `None` — this is exactly what `axum_extra::extract::Query` gets wrong
    /// and this function exists to route around.
    #[test]
    fn raw_present_but_empty_is_some_empty_string_not_none() {
        assert_eq!(
            raw_environment_id(Some("environment_id=")),
            Some(String::new())
        );
        assert_eq!(
            raw_environment_id(Some("since_days=7&environment_id=&limit=20")),
            Some(String::new())
        );
    }

    #[test]
    fn raw_present_with_value_is_decoded() {
        let id = Uuid::from_u128(7);
        assert_eq!(
            raw_environment_id(Some(&format!("environment_id={id}"))),
            Some(id.to_string())
        );
    }
}
